File: 活動選擇問題/活動選擇問題.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer st = new StreamTokenizer(br);
        ArrayList<Activity> activities = new ArrayList<>();
        while(st.nextToken() == StreamTokenizer.TT_NUMBER) {
            int start = (int)st.nval;
            st.nextToken();
            int end = (int)st.nval;
            if(start == 0 && end == 0)
                break;
            activities.add(new Activity(start, end));
        }
        activities.sort(new Comparator<Activity>() {
            @Override
            public int compare(Activity a1, Activity a2) {
                return Integer.compare(a1.end, a2.end);
            }
        });
        int count = 0, last = 0;
        for(Activity a : activities) {
            if(a.start >= last) {
                last = a.end;
                ++count;
            }
        }
        System.out.println("Maximum number of non-overlapping activities: " + count);
	}
}

class Activity {
    int start, end;
    Activity(int start, int end) {
        this.start = start;
        this.end = end;
    }
}

File: 活動選擇問題/活動選擇問題.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int start, end;
} Activity;

int compare(const void *a, const void *b) {
    return ((Activity*)a)->end > ((Activity*)b)->end;
}

int main() {
    Activity *activities = NULL;
    int size = 0, start, end;
    while(scanf("%d %d", &start, &end) && (start != 0 || end != 0)) {
        activities = (Activity*)realloc(activities, (size + 1) * sizeof(Activity));
        activities[size].start = start;
        activities[size++].end = end;
    }
    qsort(activities, size, sizeof(Activity), compare);
    int count = 0, last = 0;
    for(int i = 0; i < size; ++i) {
        if(activities[i].start >= last) {
            last = activities[i].end;
            ++count;
        }
    }
    printf("Maximum number of non-overlapping activities: %d\n", count);
    return 0;
}
